karapace/schema_models.py

```diff
@@ -51,41 +55,56 @@ def parse_jsonschema_definition(schema_definition: str) -> Draft7Validator:
     return Draft7Validator(schema)
 
 
-def parse_protobuf_schema_definition(schema_definition: str) -> ProtobufSchema:
+def parse_protobuf_schema_definition(
+    schema_definition: str,
+    references: Optional[List[Reference]] = None,
+    dependencies: Optional[Dict[str, Dependency]] = None,
+    validate_references: bool = True,
+) -> ProtobufSchema:
     """Parses and validates `schema_definition`.
 
     Raises:
-        Nothing yet.
+        ProtobufUnresolvedDependencyException if Protobuf dependency cannot be resolved.
 
     """
-
-    return ProtobufSchema(schema_definition)
-
-
-@unique
-class SchemaType(str, Enum):
-    AVRO = "AVRO"
-    JSONSCHEMA = "JSON"
-    PROTOBUF = "PROTOBUF"
+    protobuf_schema = ProtobufSchema(schema_definition, references, dependencies)
+    if validate_references:
+        result = protobuf_schema.verify_schema_dependencies()
+        if not result.result:
+            raise ProtobufUnresolvedDependencyException(f"{result.message}")
+    return protobuf_schema
 
 
 def _assert_never(no_return: NoReturn) -> NoReturn:
     raise AssertionError(f"Expected to be unreachable {no_return}")
 
 
 class TypedSchema:
-    def __init__(self, schema_type: SchemaType, schema_str: str):
+    def __init__(
+        self,
+        *,
+        schema_type: SchemaType,
+        schema_str: str,
+        schema: Optional[Union[Draft7Validator, AvroSchema, ProtobufSchema]] = None,
+        references: Optional[List[Reference]] = None,
+        dependencies: Optional[Dict[str, Dependency]] = None,
+    ):
         """Schema with type information
 
         Args:
             schema_type (SchemaType): The type of the schema
             schema_str (str): The original schema string
+            schema (Optional[Union[Draft7Validator, AvroSchema, ProtobufSchema]]): The parsed and validated schema
+             references (Optional[List[Dependency]]): The references of schema
         """
         self.schema_type = schema_type
-        self.schema_str = TypedSchema.normalize_schema_str(schema_str, schema_type)
+        self.references = references
+        self.dependencies = dependencies
+        self.schema_str = TypedSchema.normalize_schema_str(schema_str, schema_type, references, dependencies)
         self.max_id: Optional[SchemaId] = None
-
         self._fingerprint_cached: Optional[str] = None
+        self._str_cached: Optional[str] = None
+        self._schema_cached: Optional[Union[Draft7Validator, AvroSchema, ProtobufSchema]] = schema
```

Comment:
This  is will cause heavy memory usage as `TypedSchema` is used in the in memory database and must be changed.
The usage is here: https://github.com/aiven/karapace/blob/main/karapace/schema_reader.py#L477-L489

Very early implementation also stored the schemas as objects and caused issues on low memory environments. The schema should be in object format only in the `ParsedTypedSchema` and `ValidatedTypedSchema`.

Reply:
Oh, this makes sense: even the name `ParsedTypedSchema` only makes sense if it's the one holding the schema (object).  

@libretto Did you choose to cache the schema in the `TypedSchema` just for performance reasons to avoid reparsing?  This chould be revisited as a future improvement.

Reply:
@RyanSkraba Caching was suggested to my code by @jjaakola-aiven at https://github.com/aiven/karapace/commit/984d794b73f308b762965eea3ddabbcd2e5116a3 and I follow his way. Let me try to fix it.

Reply:
Yes, it is the commit referenced. It is a mistake from me. The issue is when schema-property is called from the object used in storage that the memory use will creep in the background. I think the compatibility check would have been manifesting this problem in some point, especially transitive which iterates all previous schemas. The current implementation creates always a new `ParsedTypedSchema` object for compatibility check and caching would not even be effective: https://github.com/aiven/karapace/blob/main/karapace/schema_registry.py#L361-L363

Reply:
@jjaakola-aiven actually You were right there at this point because You also created the property 'schema'  that will utilize caching in case of multiple calls to this property. But if the schema property is called only once caching will be useless. So it was the right solution but if You know about the memory consumption issue probably we must get rid of it or improve it.

Reply:
@RyanSkraba I just moved cashing to ParsedTypedSchema hope there will be less memory issues